castle/cms/browser/controlpanel/announcements.py

```diff
@@ -220,6 +226,215 @@ def handle_text(self, action):
                 self.context.absolute_url()))
 
 
+class IExportSubscribersForm(model.Schema):
+    directives.widget('export_categories', SelectFieldWidget)
+    export_categories = schema.List(
+        title=u'Subscribers to export',
+        description=u'Leave Empty For All Subscribers',
+        required=False,
+        value_type=schema.Choice(
+            vocabulary='castle.cms.vocabularies.EmailCategories'
+        )
+    )
+
+
+class ExportSubscribersForm(AutoExtensibleForm, form.Form):
+    schema = IExportSubscribersForm
+
+    ignoreContext = True
+
+    @button.buttonAndHandler(u'Export', name='export')
+    def handle_export(self, action):
+        data, errors = self.extractData()
+        if not errors:
+            response = self.request.response
+            cd = 'attachment; filename=subscribers.csv'
+            response.setHeader('Content-Disposition', cd)
+            fields = ['name', 'email', 'phone_number', 'phone_number_confirmed',
+                      'confirmed', 'code', 'created', 'captcha', 'categories']
+            responsebody = ','.join(fields)
+            categories = set()
+            if 'form.widgets.export_categories' in self.request.form:
+                if self.request.form['form.widgets.export_categories'] != u'':
```

Comment:
I'm curious, aren't these values available on the `data` object? Maybe `data['export_categories']` instead of `self.request.form['form.widgets.export_categories']`. It should also deserialize the value for you too.

Reply:
I stole it from [here](https://github.com/castlecms/castle.cms/blob/master/castle/cms/browser/controlpanel/announcements.py#L186). Looks way nicer (and makes more sense) using data. Commit incoming.